src/lib/components/EventSchedule/index.js

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import '../EventScheduleModal';
+
+/**
+ * @fileoverview A schedule manager which opens and closes modals based on
+ * the current href. Looks in children for content.
+ *
+ * This does not inherit from BaseStateElement as it is not a LitElement.
+ */
+
+class EventSchedule extends HTMLElement {
+  constructor() {
+    super();
+
+    this._modalElement = document.createElement('web-event-schedule-modal');
+    this._modalElement.className = 'web-modal';
+    this._modalElement.open = false;
+    this._modalElement.addEventListener('close-modal', () => {
+      if (!window.location.hash.substr(1)) {
+        return; // we closed ourselves, so got an event that can be ignored
+      }
+      const url = window.location.pathname + window.location.search;
+      window.history.replaceState(null, null, url);
+      this.onHashChange();
+    });
+
+    this._currentSession = null;
+    this.onHashChange = this.onHashChange.bind(this);
+    this.onClick = this.onClick.bind(this);
+  }
+
+  _elementForHash(hash) {
+    const id = hash.substr(1);
+    return (id && this.querySelector(`[data-session-id="${id}"]`)) || null;
+  }
+
+  /**
+   * Controls the open/close of session modals, by matching the page's hash to any contained
+   * elements with a matching `data-session-id` attribute.
+   *
+   * If no hash is set, or no element is found, hides any modal.
+   */
+  onHashChange() {
+    const session = this.isConnected
+      ? this._elementForHash(window.location.hash)
+      : null;
+    if (session === this._currentSession) {
+      return;
+    }
+
+    this._currentSession = session;
+    if (!session) {
+      this._modalElement.open = false;
+      this._modalElement.remove();
+      return;
+    }
+
+    // Clone the session node and pass it to our session. This is kinda gross but basically we use
+    // it as the canonical source of truth for the modal. We also have to remove all tabindex
+    // attributes as they may have been added by the inert polyfill.
+    const clone = session.cloneNode(true);
```

Comment:
I think this implementation is totally fine, but as I was working on the livestream player I realized I was hitting a similar problem. Basically we don't have an API to fetch data, and we're not bootstrapping the page with data. So we're doing workarounds.

For the measure page we serialize a data object and add it to the page. This is how the measure tool knows how to link an audit to a specific guide. I was curious if we should consider using that approach more often in the future.

Reply:
I agree but also... I think at this point I'm happy for bespoke things for this event.